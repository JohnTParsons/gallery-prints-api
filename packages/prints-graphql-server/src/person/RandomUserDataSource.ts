/* This is just here as an example of a different API to merge into the combined GraphQL API */

import { RESTDataSource } from 'apollo-datasource-rest';

export class RandomUserDataSource extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = 'https://api.randomuser.me/';
  }

  async getPerson() {
    const { results } = await this.get('');
    return results;
  }
}
