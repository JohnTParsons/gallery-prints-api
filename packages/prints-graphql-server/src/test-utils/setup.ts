import { RESTDataSource } from 'apollo-datasource-rest';
import { InMemoryLRUCache } from 'apollo-server-caching';

/**
 * Initialize the service.
 * NOTE:
 * ==================================
 * We need to initialize service (dataSource) manually as if it's used through apollo server, apollo server also do the same.
 * ==================================
 * @param Service
 */
export const initialize = <T extends RESTDataSource>(Service: new () => T): T => {
  const service = new Service();

  service.initialize({
    context: {},
    cache: new InMemoryLRUCache(),
  });
  return service;
};
