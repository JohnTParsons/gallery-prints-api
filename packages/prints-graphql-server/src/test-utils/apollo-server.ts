import { createTestClient } from 'apollo-server-testing';
//import schema from '../schema'; // import existing schema & datasource to utilize the same in test scenarios as well...
import { ApolloServer, gql } from 'apollo-server';
import { RandomUserDataSource } from '../person/RandomUserDataSource';
import { HarvardArtPrintDataSource } from '../print/HarvardArtPrintDataSource';
import * as commonTypes from '../common/common-types';
import * as personTypes from '../person/person-types';
import * as printTypes from '../print/print-types';
import * as personResolvers from '../person/person-resolvers';
import * as printResolvers from '../print/print-resolvers';

const typeDefs = gql`
  ${commonTypes.typeDefString}
  ${personTypes.typeDefString}
  ${printTypes.typeDefString}
  type Query {
    ${personTypes.queryTypeDefString}
    ${printTypes.queryTypeDefString}
  }
`;

const resolvers = {
  Query: {
    ...personResolvers.resolvers.Query,
    ...printResolvers.resolvers.Query,
  },
};

const createApolloTestServer = () => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      randomUserAPI: new RandomUserDataSource(),
      harvardArtPrintAPI: new HarvardArtPrintDataSource(),
    }),
    debug: true,
    context: () => {
      // Important: request context is not available during jest testing, hence mocked the request context here.
      return { token: 'foo' };
    },
  });
  return server;
};

const apolloServer = createApolloTestServer();

// @ts-ignore
export const { query, mutate } = createTestClient(apolloServer);
