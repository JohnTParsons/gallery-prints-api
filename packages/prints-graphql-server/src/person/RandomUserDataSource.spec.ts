import nock from 'nock';
import { initialize } from '../test-utils/setup';
import { RandomUserDataSource } from './RandomUserDataSource';

const validRestResponse = { results: { foo: 'bar' } };

describe('RandomUserDataSource', () => {
  let randomUserAPI: RandomUserDataSource;

  beforeAll(() => {
    randomUserAPI = initialize(RandomUserDataSource);
  });

  it('should return person object', async () => {
    nock(
      'https://api.randomuser.me',
      // Wildcard match the path for simplicity
      {
        filteringScope: function (scope) {
          return true;
        },
      },
    )
      .filteringPath(function (path) {
        return '/';
      })
      .get('/')
      .reply(200, validRestResponse);
    const response = await randomUserAPI.getPerson();
    expect(response).toEqual({ foo: 'bar' });
    expect(response).toMatchSnapshot();
  });
});
