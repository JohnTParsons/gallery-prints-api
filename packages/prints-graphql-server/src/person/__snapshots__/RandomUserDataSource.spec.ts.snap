// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomUserDataSource should return person object 1`] = `
Object {
  "foo": "bar",
}
`;
